# Prepping Migrate input files on Hyak

#### M. Fisher 
#### 2021-08-25


## Description
Prepare Migrate input files using Stacks output. 

(1) Access the MERlab Hyak node, (2) download Stacks v1.44, and (3) re-run `populations` to produce a RAD loci FASTA file. 

Resources: [Nam Pho's Tutorial](https://github.com/merlab-uw/Klone/blob/main/HowToUseKlone.md)

<br>
<br>

## 1: Access the MERlab Hyak Node

This is my first time entering Hyak, and I'm doing it remotely from home. First, I connected to the UW Campus Traffic Network using the Big-IP client. Then, from the **git bash** terminal:

```
ssh mfisher5@klone.hyak.uw.edu    # login with UW net ID

> ECDSA key fingerprint is <>.
> Are you sure you want to continue connecting (yes/no)? 
yes
> Warning: Permanently added 'klone.hyak.uw.edu' to the list of known hosts.
> Password:

my-UW-password                     # enter UW net ID password

> Duo two-factor login for mfisher5
>
> Enter a passcode or select one of the following options:
>
>  1. Duo Push
>  2. Phone call
>
> Passcode or option (1-2):

2                                   # opt for phone call 


cd /gscratch/merlab/                # go to MERlab's home directory
ls
> ctarpey  elpetrou  genomes  mfisher  nclowell  pollock_wgs  singularity_sif  software  usage_report.txt

cd mfisher
ls
> stacks_b8_wgenome
```

Carolyn copied in all of my intermediate Stacks files that I'll need to run populations. 


## 2: Download and Install Stacks

I went looking for older versions of Stacks, and found that the following are available as downloads on the [*Frequently Asked Questions*](https://catchenlab.life.illinois.edu/stacks/faq.php#prev) webpage: 

- stacks-1.19.tar.gz
- stacks-1.23.tar.gz
- stacks-1.27.tar.gz
- stacks-1.29.tar.gz
- stacks-1.37.tar.gz
- This is the final version 1 release: stacks-1.48.tar.gz

v1.48 is the closest I can get to v1.44, so I right clicked on the download link and selected *Inspect*. From there, I got this download link: 

`<a href="http://catchenlab.life.illinois.edu/stacks/source/stacks-1.48.tar.gz">stacks-1.48.tar.gz</a>`

Because I cannot edit files in a directory that belongs to a different Hyak user, I created a new root directory for myself (mfisher5). After navigating into that directory, I ran the following code to download and install Stacks v1.48. 

```
wget http://catchenlab.life.illinois.edu/stacks/source/stacks-1.48.tar.gz
tar xzvf stacks-1.48.tar.gz
cd stacks-1.48 
./configure
make 
```

## 3: Re-run `Populations` (or not)

The [`populations` log](https://github.com/mfisher5/Pacific-cod-RADseq/blob/main/data/stacks/batch_7.populations.log) from my final run of the program (2017-09-28) recorded the following input code: 

```
populations -b 7 -P stacks_b8_wgenome -M scripts/PopMap_L1-5.txt -t 36 -r 0.80 -p 4 -m 10 --write_random_snp --genepop --fasta
```

The new documentation for Stacks reads: 

> The per-locus, per-haplotpye FASTA output from the populations program, generated by specifying the --fasta-samples option, provides the full sequence from the RAD locus, for each haplotype, from every sample in the population, for each catalog locus. This output is identical to the populations.haplotypes.tsv output, except it provides the variable sites embedded within the full consensus seqeunce for the locus.

I did not use the `--fasta-samples` option in my original run. 

But there is this old documentation that sounds like the `--fasta` option in my original run may have produced the output I need:

![stacks-doc-img](https://github.com/mfisher5/Pacific-cod-RADseq/blob/main/doc/imgs/stacks-old-populations-fasta-doc.png?raw=true)

To check this out, I'll pull my file `batch_7.fa` off of Hyak. 

```
rsync --verbose --archive --progress mfisher5@klone.hyak.uw.edu:/gscratch/merlab/mfisher/stacks_b8_wgenome/batch_7.fa /Downloads
```
<br>

The Stacks manual then provides this code to get the internal Sample IDs used by Stacks:
```
ls -1 *.alleles.tsv | 
  while read line; do echo -n $line | 
  sed -E -e 's/^(.+)\.alleles\.tsv$/\1\t/'; cat $line | head -n 1 | cut -f 2;
  done
  
>BOR07_01        # pstacks version 1.44; generated on 2017-09-28 04:05:27
>BOR07_01_2      # pstacks version 1.44; generated on 2017-09-28 04:24:22
>BOR07_02.1      # pstacks version 1.44; generated on 2017-09-28 04:06:02
>BOR07_03        # pstacks version 1.44; generated on 2017-09-28 04:06:36
>BOR07_04        # pstacks version 1.44; generated on 2017-09-28 04:07:28
>BOR07_05.1      # pstacks version 1.44; generated on 2017-09-28 04:08:41
>BOR07_06.1      # pstacks version 1.44; generated on 2017-09-28 04:09:34
>BOR07_07.1      # pstacks version 1.44; generated on 2017-09-28 04:10:24

```

I have to change it a bit to read my old files....

```
ls -1 *.alleles.tsv | 
  while read line; do echo -n $line | 
  sed -E -e 's/^(.+)\.alleles\.tsv$/\1\t/'; cat $line | head -n 2 | tail -n 1 | cut -f 2;
  done
  
>BOR07_01        267
>BOR07_01_2      289
>BOR07_02.1      268
>BOR07_03        269
>BOR07_04        270
>BOR07_05.1      271
>BOR07_06.1      272
```

This worked for every file except GEO020414_13.alleles.tsv, which is apparently empty...
I copied the output in the terminal into a text file, that I can use to interpret the fasta file. 


## 4: Generate Migrate input files

I can maybe use a python script by Paul Maier (April 22, 2016; Written for Python 2.7.5) to generate the Migrate files from the populations fasta file. 
I need to bring some files into Hyak to do this:

```
# attempt 1
rsync --verbose --archive --progress /Documents/Pacific-cod-RADseq/data/migrate/migrate_loci_list.txt mfisher5@klone.hyak.uw.edu:/gscratch/merlab/mfisher5/migrate
# attempt 2
rsync --verbose --archive --progress /Documents/migrate_loci_list.txt mfisher5@klone.hyak.uw.edu:/gscratch/merlab/mfisher5/migrate
# attempt 3
rsync --verbose --archive --progress /c/Users/mcf05/Documents/Pacific-cod-RADseq/data/migrate/migrate_loci_list.txt mfisher5@klone.hyak.uw.edu:/gscratch/merlab/mfisher5/migrate
# attempt 4
scp /Documents/migrate_loci_list.txt mfisher5@klone.hyak.uw.edu:/gscratch/merlab/mfisher5/migrate
# attempt 5
scp /c/Users/mcf05/Documents/migrate_loci_list.txt mfisher5@klone.hyak.uw.edu:/gscratch/merlab/mfisher5/migrate
# attempt 6
scp ./migrate_loci_list.txt mfisher5@klone.hyak.uw.edu:/gscratch/merlab/mfisher5/migrate
```

And install Python 2.7.5
```
wget https://www.python.org/ftp/python/2.7.5/Python-2.7.5.tgz
tar xzvf Python-2.7.5.tgz
cd Python-2.7.5
./configure
make 

python2 -V
>Python 2.7.18
```


Then I *should* be able to run the following code from within my `mfisher5` folder:
``` 
python2 migrate/fasta2genotype.py migrate/batch_7.fa migratemigrate_loci_list.txt migrate/migrate_stacksIDs_finalgenepop NA migrate/pcod-korea-migrate-input
```

----

(take 2)

<br>


## 5 Rerun `Populations` (for real)

The `fasta2genotype.py` script does not seem to be working, so I am going to re-run Stacks' `populations` on Hyak and produce a VCF file. Then I'll replicate the filtering (for individuals, loci) that I did after the original `populations` run. This is based on MerLab's [How to use Slurm](https://github.com/merlab-uw/Klone/blob/main/HowToUseSlurm.md) documentation. 

Returning a VCF file from `populations` requires a minor change to my original input code: 
```
populations -b 7 -P stacks_b8_wgenome -M scripts/PopMap_L1-5.txt -t 36 -r 0.80 -p 4 -m 10 --write_random_snp --genepop --fasta --vcf
```

I think that to use `slurm` on Hyak, I need this code saved into a [.sh file](https://github.com/mfisher5/Pacific-cod-RADseq/blob/main/scripts/rerun_populations.sh). 


The full code for re-running populations on Hyak:

```
# navigate to home directory
cd /gscratch/merlab/mfisher5   

# create stacks subfolder
mkdir stacks_b8_wgenome

# copy over stacks intermediate files
cp -a /gscratch/merlab/mfisher/stacks_b8_wgenome/. /gscratch/merlab/mfisher5/stacks_b8_wgenome

# upload population map (note - started Hyak while in 'Documents/Pacific-cod-RADseq' directory)
rsync --verbose --archive --progress /Documents/Pacific-cod-RADseq/data/stacks/PopMap_L1-5.txt mfisher5@klone.hyak.uw.edu:/gscratch/merlab/mfisher5/stacks_b8_wgenome
# error: rsync: change_dir "/Documents/Pacific-cod-RADseq/data/stacks" failed: No such file or directory (2)

rsync --verbose --archive --progress ~/Documents/Pacific-cod-RADseq/data/stacks/PopMap_L1-5.txt mfisher5@klone.hyak.uw.edu:/gscratch/merlab/mfisher5/stacks_b8_wgenome
# error: rsync: change_dir "/mmfs1/home/mfisher5/Documents/Pacific-cod-RADseq/data/stacks" failed: No such file or directory (2)

rsync --verbose --archive --progress ./data/stacks/PopMap_L1-5.txt mfisher5@klone.hyak.uw.edu:/gscratch/merlab/mfisher5/stacks_b8_wgenome
# error: rsync: change_dir "/mmfs1/gscratch/merlab/mfisher5//./data/stacks" failed: No such file or directory (2)


rsync --verbose --archive --progress /data/stacks/PopMap_L1-5.txt mfisher5@klone.hyak.uw.edu:/gscratch/merlab/mfisher5/stacks_b8_wgenome
# error: rsync: change_dir "/data/stacks" failed: No such file or directory (2)

rsync --verbose --archive --progress /c/Users/mcf05/Documents/Pacific-cod-RADseq/data/stacks/PopMap_L1-5.txt mfisher5@klone.hyak.uw.edu:/gscratch/merlab/mfisher5/stacks_b8_wgenome
# error: rsync: change_dir "/c/Users/mcf05/Documents/Pacific-cod-RADseq/data/stacks" failed: No such file or directory (2)


# upload populations batch file
rsync --verbose --archive --progress /Documents/Pacific-cod-RADseq/data/scripts/rerun_populations.sh mfisher5@klone.hyak.uw.edu:/gscratch/merlab/mfisher5

# rerun populations (assumes 'stacks' subdirectory contains stacks intermediate files)
sbatch rerun_populations.sh

```

If I can't get this to work, I think Carolyn may have to run it out of the 'mfisher' directory (adjusted .sh file [here](https://github.com/mfisher5/Pacific-cod-RADseq/blob/main/scripts/rerun_populations_carolyn.sh) ).

```
# navigate to home directory
cd /gscratch/merlab/mfisher   

# upload population map 
rsync --verbose --archive --progress /Pacific-cod-RADseq/data/stacks/PopMap_L1-5.txt ctarpey@klone.hyak.uw.edu:/gscratch/merlab/mfisher/stacks_b8_wgenome

# upload populations batch file 
rsync --verbose --archive --progress /Pacific-cod-RADseq/data/scripts/rerun_populations_carolyn.sh ctarpey@klone.hyak.uw.edu:/gscratch/merlab/mfisher


# rerun populations (assumes 'stacks' subdirectory contains stacks intermediate files)
sbatch rerun_populations.sh

# see information about the job and the job id
squeue --account merlab

```
## Running Stacks on Hyak for a VCF output

Mary wrote the code that we needed to run Stacks on Hyak. I used it as the starting point for getting Stacks to run on Hyak with the input files that we have. 

rsync to move the code up to Hyak, as well as the PopMap_L1-5.txt, the last part of the input files we needed. 
```
ctarpey@DESKTOP-CG1JP0S:~$ rsync --verbose --archive --progress /mnt/d/PacificCodWGS/Mary_PcodAnnotations/stacks/rerun_populations_carolyn.sh  ctarpey@klone.hyak.uw.edu:/gscratch/merlab/mfisher/

ctarpey@DESKTOP-CG1JP0S:~$ rsync --verbose --archive --progress /mnt/d/PacificCodWGS/Mary_PcodAnnotations/stacks/PopMap_L1-5.txt ctarpey@klone.hyak.uw.edu:/gscratch/merlab/mfisher/stacks_b8_wgenome/
```

This is the shell script to sbatch file that I used to run Stacks on Hyak rerun_populations_carolyn.sh
```
#!/bin/bash
#SBATCH --job-name=mcf_stacks
#SBATCH --account=merlab
#SBATCH --partition=compute-hugemem
#SBATCH --nodes=1
## Walltime (days-hours:minutes:seconds format)
#SBATCH --time=3-12:00:00
## Memory per node
#SBATCH --mem=100G
#SBATCH --mail-type=ALL
#SBATCH --mail-user=ctarpey@uw.edu

## CODE FOR JOB

/gscratch/merlab/mfisher5/stacks-1.48/populations -b 7 -P /gscratch/merlab/mfisher/stacks_b8_wgenome -M /gscratch/merlab/mfisher/stacks_b8_wgenome/PopMap_L1-5.txt -t 36 -r 0.80 -p 4 -m 10 --write_random_snp --genepop --fasta --vcf
```

Command I used to start the job, and the job batch number
```
(base) [ctarpey@klone1 mfisher]$ sbatch -A merlab rerun_populations_carolyn.sh
Submitted batch job 929388
```

Its running! 
```
(base) [ctarpey@klone1 mfisher]$ squeue -A merlab
             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)
            929383 compute-h herring_ elpetrou  R       2:52      1 n3302
            929388 compute-h mcf_stac  ctarpey  R       0:13      1 n3017
            928369 compute-h pollock_ elpetrou  R    1:49:42      1 n3017
```

I took a look at the slurm output file, which has what would have been printed to the terminal if this were run in a typical terminal on a desktop. It appears to be running with no errors. 
```
(base) [ctarpey@klone1 mfisher]$ cat slurm-929388.out
populations parameters selected:
  Fst kernel smoothing: off
  Bootstrap resampling: off
  Percent samples limit per population: 0.8
  Locus Population limit: 4
  Minimum stack depth: 10
  Log liklihood filtering: off; threshold: 0
  Minor allele frequency cutoff: 0
  Maximum observed heterozygosity cutoff: 1
  Applying Fst correction: none.

Parsing population map...
The population map contained 326 samples, 9 population(s), 1 group(s).
Reading the catalog...
  Parsing /gscratch/merlab/mfisher/stacks_b8_wgenome/batch_7.catalog.tags.tsv
  Parsing /gscratch/merlab/mfisher/stacks_b8_wgenome/batch_7.catalog.snps.tsv
  Parsing /gscratch/merlab/mfisher/stacks_b8_wgenome/batch_7.catalog.alleles.tsv
Reading matches to the catalog...
  Parsing /gscratch/merlab/mfisher/stacks_b8_wgenome/PO010715_02.matches.tsv
  Parsing /gscratch/merlab/mfisher/stacks_b8_wgenome/PO010715_02_rep.matches.tsv
  Parsing /gscratch/merlab/mfisher/stacks_b8_wgenome/PO010715_04.matches.tsv
  Parsing /gscratch/merlab/mfisher/stacks_b8_wgenome/PO010715_06.1.matches.tsv
  Parsing /gscratch/merlab/mfisher/stacks_b8_wgenome/PO010715_07.matches.tsv
  Parsing /gscratch/merlab/mfisher/stacks_b8_wgenome/PO010715_07_rep.matches.tsv
  Parsing /gscratch/merlab/mfisher/stacks_b8_wgenome/PO010715_08.1.matches.tsv
  Parsing /gscratch/merlab/mfisher/stacks_b8_wgenome/PO010715_10.1.matches.tsv
  Parsing /gscratch/merlab/mfisher/stacks_b8_wgenome/PO010715_11.1.matches.tsv
  Parsing /gscratch/merlab/mfisher/stacks_b8_wgenome/PO010715_12.matches.tsv
  Parsing /gscratch/merlab/mfisher/stacks_b8_wgenome/PO010715_17.1.matches.tsv
{........}
Regenerating nucleotide-level summary statistics for population 'Jukbyeon07'
Population 'Jukbyeon07' contained 0 incompatible loci -- more than two alleles present.
Regenerating nucleotide-level summary statistics for population 'JinhaeBay07'
Population 'JinhaeBay07' contained 0 incompatible loci -- more than two alleles present.
Regenerating nucleotide-level summary statistics for population 'JinhaeBay08'
Population 'JinhaeBay08' contained 0 incompatible loci -- more than two alleles present.
Regenerating nucleotide-level summary statistics for population 'Boryeong07'
Population 'Boryeong07' contained 0 incompatible loci -- more than two alleles present.
Regenerating nucleotide-level summary statistics for population 'Geoje14'
Population 'Geoje14' contained 0 incompatible loci -- more than two alleles present.
Re-tallying loci across populations...done.
Generating haplotype-level summary statistics for population 'Pohang15'
Generating haplotype-level summary statistics for population 'Geoje15'
Generating haplotype-level summary statistics for population 'Namhae15'
Generating haplotype-level summary statistics for population 'YellowSea16'
Generating haplotype-level summary statistics for population 'Jukbyeon07'
Generating haplotype-level summary statistics for population 'JinhaeBay07'
Generating haplotype-level summary statistics for population 'JinhaeBay08'
Generating haplotype-level summary statistics for population 'Boryeong07'
Generating haplotype-level summary statistics for population 'Geoje14'
Writing 22807 loci to summary statistics file, '/gscratch/merlab/mfisher/stacks_b8_wgenome/batch_7.sumstats.tsv'
Writing 22807 loci to observed haplotype file, '/gscratch/merlab/mfisher/stacks_b8_wgenome/batch_7.haplotypes.tsv'
Writing sample alleles (raw) to FASTA file '/gscratch/merlab/mfisher/stacks_b8_wgenome/batch_7.samples-raw.fa'
Writing population data to GenePop file '/gscratch/merlab/mfisher/stacks_b8_wgenome/batch_7.genepop'
Writing population data to VCF file '/gscratch/merlab/mfisher/stacks_b8_wgenome/batch_7.vcf'
Populations is done.
```

It did output a VCF file that looks totally normal. 

## Filtering the VCF file from Stacks 

With the new VCF file from Stacks, we have to filter it to retain only the loci and individuals that we are interested in using in the Migrate analysis. I also have to rename the remaining individuals so that their names are 9 characters or less, a requirement of Migrate. 

The list of loci that we want to use is located in one of the very first files that Mary shared with me. Its called migrate_loci_list.txt. The individuals that we want to keep is with that file and is called migrate_stacksIDs_finalgenepop.txt. 

I'll use those files to filter the VCF with VCFtools. 
```
ctarpey@DESKTOP-CG1JP0S:~$ vcftools

VCFtools (0.1.16)
Â© Adam Auton and Anthony Marcketta 2009
```

From the manual, these are the command options I am interested in: 

```
vcftools [ --vcf FILE | --gzvcf FILE | --bcf FILE] [ --out OUTPUT PREFIX ] [ FILTERING OPTIONS ] [ OUTPUT OPTIONS ]

--keep <filename>
--remove <filename>

Provide files containing a list of individuals to either include or exclude in subsequent analysis. Each individual ID (as defined in the VCF headerline) should be included on a separate line. If both options are used, then the "--keep" option is executed before the "--remove" option. When multiple files are provided, the union of individuals from all keep files subtracted by the union of individuals from all remove files are kept. No header line is expected.

--snps <filename>
--exclude <filename>

Include or exclude a list of SNPs given in a file. The file should contain a list of SNP IDs (e.g. dbSNP rsIDs), with one ID per line. No header line is expected.

```

From that I gather this is the command I want to run to retain only the individuals in the migrate_stacksIDs_finalgenepop.txt and the loci in the migrate_loci_list.txt. 

```
vcftools  --vcf /mnt/d/PacificCodWGS/Mary_PcodAnnotations/Migrate/Stacks_onHyak_forVCF/batch_7.vcf --out batch_7_filteredInds_loci --keep /mnt/d/PacificCodWGS/Mary_PcodAnnotations/Migrate/FilterVCF/migrate_stacksIDs_finalgenepop.txt --snps /mnt/d/PacificCodWGS/Mary_PcodAnnotations/Migrate/FilterVCF/migrate_loci_list.txt 
```

That didn't work 
```
Parameters as interpreted:
        --vcf /mnt/d/PacificCodWGS/Mary_PcodAnnotations/Migrate/Stacks_onHyak_forVCF/batch_7.vcf
        --keep /mnt/d/PacificCodWGS/Mary_PcodAnnotations/Migrate/FilterVCF/migrate_stacksIDs_finalgenepop.txt
        --out batch_7_filteredInds_loci
        --snps /mnt/d/PacificCodWGS/Mary_PcodAnnotations/Migrate/FilterVCF/migrate_loci_list.txt

Keeping individuals in 'keep' list
After filtering, kept 0 out of 326 Individuals
After filtering, kept 0 out of a possible 21128 Sites
No data left for analysis!
Run Time = 1.00 seconds
```

I needed to make some changes to the migrate_stacksIDs_finalgenepop.txt file, it had three columns and a header. I removed the header and retained only the column with the name of the sample. I saved the new file as migrate_stacksIDs_finalgenepop_namesOnly.txt

```
vcftools  --vcf /mnt/d/PacificCodWGS/Mary_PcodAnnotations/Migrate/Stacks_onHyak_forVCF/batch_7.vcf --out batch_7_filteredInds_loci --keep /mnt/d/PacificCodWGS/Mary_PcodAnnotations/Migrate/FilterVCF/
migrate_stacksIDs_finalgenepop_namesOnly.txt --snps /mnt/d/PacificCodWGS/Mary_PcodAnnotations/Migrate/FilterVCF/migrate_loci_list.txt
```
