# Prepping Migrate input files on Hyak

#### M. Fisher 
#### 2021-08-25


## Description
Prepare Migrate input files using Stacks output. 

(1) Access the MERlab Hyak node, (2) download Stacks v1.44, and (3) re-run `populations` to produce a RAD loci FASTA file. 

Resources: [Nam Pho's Tutorial](https://github.com/merlab-uw/Klone/blob/main/HowToUseKlone.md)

<br>
<br>

## 1: Access the MERlab Hyak Node

This is my first time entering Hyak, and I'm doing it remotely from home. First, I connected to the UW Campus Traffic Network using the Big-IP client. Then, from the **git bash** terminal:

```
ssh mfisher5@klone.hyak.uw.edu    # login with UW net ID

> ECDSA key fingerprint is <>.
> Are you sure you want to continue connecting (yes/no)? 
yes
> Warning: Permanently added 'klone.hyak.uw.edu' to the list of known hosts.
> Password:

my-UW-password                     # enter UW net ID password

> Duo two-factor login for mfisher5
>
> Enter a passcode or select one of the following options:
>
>  1. Duo Push
>  2. Phone call
>
> Passcode or option (1-2):

2                                   # opt for phone call 


cd /gscratch/merlab/                # go to MERlab's home directory
ls
> ctarpey  elpetrou  genomes  mfisher  nclowell  pollock_wgs  singularity_sif  software  usage_report.txt

cd mfisher
ls
> stacks_b8_wgenome
```

Carolyn copied in all of my intermediate Stacks files that I'll need to run populations. 


## 2: Download and Install Stacks

I went looking for older versions of Stacks, and found that the following are available as downloads on the [*Frequently Asked Questions*](https://catchenlab.life.illinois.edu/stacks/faq.php#prev) webpage: 

- stacks-1.19.tar.gz
- stacks-1.23.tar.gz
- stacks-1.27.tar.gz
- stacks-1.29.tar.gz
- stacks-1.37.tar.gz
- This is the final version 1 release: stacks-1.48.tar.gz

v1.48 is the closest I can get to v1.44, so I right clicked on the download link and selected *Inspect*. From there, I got this download link: 

`<a href="http://catchenlab.life.illinois.edu/stacks/source/stacks-1.48.tar.gz">stacks-1.48.tar.gz</a>`

Because I cannot edit files in a directory that belongs to a different Hyak user, I created a new root directory for myself (mfisher5). After navigating into that directory, I ran the following code to download and install Stacks v1.48. 

```
wget http://catchenlab.life.illinois.edu/stacks/source/stacks-1.48.tar.gz
tar xzvf stacks-1.48.tar.gz
cd stacks-1.48 
./configure
make 
```

## 3: Re-run `Populations` (or not)

The [`populations` log](https://github.com/mfisher5/Pacific-cod-RADseq/blob/main/data/stacks/batch_7.populations.log) from my final run of the program (2017-09-28) recorded the following input code: 

```
populations -b 7 -P stacks_b8_wgenome -M scripts/PopMap_L1-5.txt -t 36 -r 0.80 -p 4 -m 10 --write_random_snp --genepop --fasta
```

The new documentation for Stacks reads: 

> The per-locus, per-haplotpye FASTA output from the populations program, generated by specifying the --fasta-samples option, provides the full sequence from the RAD locus, for each haplotype, from every sample in the population, for each catalog locus. This output is identical to the populations.haplotypes.tsv output, except it provides the variable sites embedded within the full consensus seqeunce for the locus.

I did not use the `--fasta-samples` option in my original run. 

But there is this old documentation that sounds like the `--fasta` option in my original run may have produced the output I need:

![stacks-doc-img](https://github.com/mfisher5/Pacific-cod-RADseq/blob/main/doc/imgs/stacks-old-populations-fasta-doc.png?raw=true)

To check this out, I'll pull my file `batch_7.fa` off of Hyak. 

```
rsync --verbose --archive --progress mfisher5@klone.hyak.uw.edu:/gscratch/merlab/mfisher/stacks_b8_wgenome/batch_7.fa /Downloads
```
<br>

The Stacks manual then provides this code to get the internal Sample IDs used by Stacks:
```
ls -1 *.alleles.tsv | 
  while read line; do echo -n $line | 
  sed -E -e 's/^(.+)\.alleles\.tsv$/\1\t/'; cat $line | head -n 1 | cut -f 2;
  done
  
>BOR07_01        # pstacks version 1.44; generated on 2017-09-28 04:05:27
>BOR07_01_2      # pstacks version 1.44; generated on 2017-09-28 04:24:22
>BOR07_02.1      # pstacks version 1.44; generated on 2017-09-28 04:06:02
>BOR07_03        # pstacks version 1.44; generated on 2017-09-28 04:06:36
>BOR07_04        # pstacks version 1.44; generated on 2017-09-28 04:07:28
>BOR07_05.1      # pstacks version 1.44; generated on 2017-09-28 04:08:41
>BOR07_06.1      # pstacks version 1.44; generated on 2017-09-28 04:09:34
>BOR07_07.1      # pstacks version 1.44; generated on 2017-09-28 04:10:24

```

I have to change it a bit to read my old files....

```
ls -1 *.alleles.tsv | 
  while read line; do echo -n $line | 
  sed -E -e 's/^(.+)\.alleles\.tsv$/\1\t/'; cat $line | head -n 2 | tail -n 1 | cut -f 2;
  done
  
>BOR07_01        267
>BOR07_01_2      289
>BOR07_02.1      268
>BOR07_03        269
>BOR07_04        270
>BOR07_05.1      271
>BOR07_06.1      272
```

This worked for every file except GEO020414_13.alleles.tsv, which is apparently empty...
I copied the output in the terminal into a text file, that I can use to interpret the fasta file. 


## 4: Generate Migrate input files

I can maybe use a python script by Paul Maier (April 22, 2016; Written for Python 2.7.5) to generate the Migrate files from the populations fasta file. 
I need to bring some files into Hyak to do this:

```
# attempt 1
rsync --verbose --archive --progress /Documents/Pacific-cod-RADseq/data/migrate/migrate_loci_list.txt mfisher5@klone.hyak.uw.edu:/gscratch/merlab/mfisher5/migrate
# attempt 2
rsync --verbose --archive --progress /Documents/migrate_loci_list.txt mfisher5@klone.hyak.uw.edu:/gscratch/merlab/mfisher5/migrate
# attempt 3
rsync --verbose --archive --progress /c/Users/mcf05/Documents/Pacific-cod-RADseq/data/migrate/migrate_loci_list.txt mfisher5@klone.hyak.uw.edu:/gscratch/merlab/mfisher5/migrate
# attempt 4
scp /Documents/migrate_loci_list.txt mfisher5@klone.hyak.uw.edu:/gscratch/merlab/mfisher5/migrate
# attempt 5
scp /c/Users/mcf05/Documents/migrate_loci_list.txt mfisher5@klone.hyak.uw.edu:/gscratch/merlab/mfisher5/migrate
# attempt 6
scp ./migrate_loci_list.txt mfisher5@klone.hyak.uw.edu:/gscratch/merlab/mfisher5/migrate
```

And install Python 2.7.5
```
wget https://www.python.org/ftp/python/2.7.5/Python-2.7.5.tgz
tar xzvf Python-2.7.5.tgz
cd Python-2.7.5
./configure
make 

python2 -V
>Python 2.7.18
```


Then I *should* be able to run the following code from within my `mfisher5` folder:
``` 
python2 migrate/fasta2genotype.py migrate/batch_7.fa migratemigrate_loci_list.txt migrate/migrate_stacksIDs_finalgenepop NA migrate/pcod-korea-migrate-input
```

----

(take 2)

<br>


## 5 Rerun `Populations` (for real)

The `fasta2genotype.py` script does not seem to be working, so I am going to re-run Stacks' `populations` on Hyak and produce a VCF file. Then I'll replicate the filtering (for individuals, loci) that I did after the original `populations` run. This is based on MerLab's [How to use Slurm](https://github.com/merlab-uw/Klone/blob/main/HowToUseSlurm.md) documentation. 

Returning a VCF file from `populations` requires a minor change to my original input code: 
```
populations -b 7 -P stacks_b8_wgenome -M scripts/PopMap_L1-5.txt -t 36 -r 0.80 -p 4 -m 10 --write_random_snp --genepop --fasta --vcf
```

I think that to use `slurm` on Hyak, I need this code saved into a [.sh file](https://github.com/mfisher5/Pacific-cod-RADseq/blob/main/scripts/rerun_populations.sh). 


The full code for re-running populations on Hyak:

```
# navigate to home directory
cd /gscratch/merlab/mfisher5   

# create stacks subfolder
mkdir stacks_b8_wgenome

# copy over stacks intermediate files
cp -a /gscratch/merlab/mfisher/stacks_b8_wgenome/. /gscratch/merlab/mfisher5/stacks_b8_wgenome

# upload population map (note - started Hyak while in 'Documents/Pacific-cod-RADseq' directory)
rsync --verbose --archive --progress /Documents/Pacific-cod-RADseq/data/stacks/PopMap_L1-5.txt mfisher5@klone.hyak.uw.edu:/gscratch/merlab/mfisher5/stacks_b8_wgenome
# error: rsync: change_dir "/Documents/Pacific-cod-RADseq/data/stacks" failed: No such file or directory (2)

rsync --verbose --archive --progress ~/Documents/Pacific-cod-RADseq/data/stacks/PopMap_L1-5.txt mfisher5@klone.hyak.uw.edu:/gscratch/merlab/mfisher5/stacks_b8_wgenome
# error: rsync: change_dir "/mmfs1/home/mfisher5/Documents/Pacific-cod-RADseq/data/stacks" failed: No such file or directory (2)

rsync --verbose --archive --progress ./data/stacks/PopMap_L1-5.txt mfisher5@klone.hyak.uw.edu:/gscratch/merlab/mfisher5/stacks_b8_wgenome
# error: rsync: change_dir "/mmfs1/gscratch/merlab/mfisher5//./data/stacks" failed: No such file or directory (2)


rsync --verbose --archive --progress /data/stacks/PopMap_L1-5.txt mfisher5@klone.hyak.uw.edu:/gscratch/merlab/mfisher5/stacks_b8_wgenome
# error: rsync: change_dir "/data/stacks" failed: No such file or directory (2)

rsync --verbose --archive --progress /c/Users/mcf05/Documents/Pacific-cod-RADseq/data/stacks/PopMap_L1-5.txt mfisher5@klone.hyak.uw.edu:/gscratch/merlab/mfisher5/stacks_b8_wgenome
# error: rsync: change_dir "/c/Users/mcf05/Documents/Pacific-cod-RADseq/data/stacks" failed: No such file or directory (2)


# upload populations batch file
rsync --verbose --archive --progress /Documents/Pacific-cod-RADseq/data/scripts/rerun_populations.sh mfisher5@klone.hyak.uw.edu:/gscratch/merlab/mfisher5

# rerun populations (assumes 'stacks' subdirectory contains stacks intermediate files)
sbatch rerun_populations.sh

```

If I can't get this to work, I think Carolyn may have to run it out of the 'mfisher' directory (adjusted .sh file [here](https://github.com/mfisher5/Pacific-cod-RADseq/blob/main/scripts/rerun_populations_carolyn.sh) ).

```
# navigate to home directory
cd /gscratch/merlab/mfisher   

# upload population map 
rsync --verbose --archive --progress /Pacific-cod-RADseq/data/stacks/PopMap_L1-5.txt ctarpey@klone.hyak.uw.edu:/gscratch/merlab/mfisher/stacks_b8_wgenome

# upload populations batch file 
rsync --verbose --archive --progress /Pacific-cod-RADseq/data/scripts/rerun_populations_carolyn.sh ctarpey@klone.hyak.uw.edu:/gscratch/merlab/mfisher


# rerun populations (assumes 'stacks' subdirectory contains stacks intermediate files)
sbatch rerun_populations.sh

# see information about the job and the job id
squeue --account merlab

```

Upload the files in the google drive file https://drive.google.com/drive/u/1/folders/1Zb8KX8gltel7AFHN29JHza8S6l68YI5w to the folder that Carolyn made on Klone for the data

```
 rsync --archive --verbose --progress /mnt/c/Users/Carolyn/Desktop/MaryFiles/* ctarpey@klone.hyak.uw.edu:/gscratch/merlab/mfisher/data/
 ```
 
Test bash script for running make_migrate_input.sh 
```
[ctarpey@klone1 data]$ cat make_migrate_inputs.sh
#!/bin/bash
#SBATCH --job-name=mfisher_test
#SBATCH --account=merlab
#SBATCH --partition=compute-hugemem
#SBATCH --nodes=1
## Walltime (days-hours:minutes:seconds format)
#SBATCH --time=3-12:00:00
## Memory per node
#SBATCH --mem=200G
#SBATCH --mail-type=ALL
#SBATCH --mail-user=ctarpey@uw.edu


## ENVIRONMENT SETUP

DATADIR=/gscratch/merlab/mfisher/data


## CODE FOR JOB

python2 $DATADIR/fasta2genotype.py $DATADIR/batch_7.fa $DATADIR/migrate_loci_list.txt $DATADIR/migrate_stacksIDs_finalgenepop.txt NA $DATADIR/pcod-korea-migrate-input
```

It is failing with this weird error 

```
[ctarpey@klone1 data]$ cat slurm-895820.out
Output type? [1] Migrate [2] Arlequin [3] DIYABC [4] LFMM [5] G-Phocs [6] Haplotype: Traceback (most recent call last):
  File "/gscratch/merlab/mfisher/data/fasta2genotype.py", line 267, in <module>
    choice = int(raw_input("Output type? [1] Migrate [2] Arlequin [3] DIYABC [4] LFMM [5] G-Phocs [6] Haplotype: "))
EOFError: EOF when reading a line
```

I used the command od -c fasta2genotype.py to see if the file has dos EOL, but it doesn't, as far as I can tell. 

``` 
ctarpey@klone1 data]$ od -c fasta2genotype.py
0000000   #   !   /   u   s   r   /   b   i   n   /   e   n   v       p
0000020   y   t   h   o   n  \n   #  \n   #       (   c   )       P   a
0000040   u   l       M   a   i   e   r  \n   #       A   p   r   i   l
0000060       2   2   ,       2   0   1   6  \n   #       f   a   s   t
0000100   a   2   g   e   n   o   t   y   p   e   .   p   y  \n   #
0000120   V       1   .   8  \n   #       W   r   i   t   t   e   n
0000140   f   o   r       P   y   t   h   o   n       2   .   7   .   5
0000160  \n   #  \n   #       T   h   i   s       p   r   o   g   r   a
0000200   m       t   a   k   e   s       a       f   a   s   t   a
0000220   f   i   l   e       l   i   s   t   i   n   g       a   l   l
0000240       a   l   l   e   l   e   s       o   f       a   l   l
0000260   i   n   d   i   v   i   d   u   a   l   s       a   t       a
0000300   l   l       l   o   c   i  \n   #       a   s       w   e   l
0000320   l       a   s       a       l   i   s   t       o   f       i
0000340   n   d   i   v   i   d   u   a   l   s   /   p   o   p   u   l
0000360   a   t   i   o   n   s       a   n   d       l   i   s   t
0000400   o   f       v   a   r   i   a   b   l   e       l   o   c   i
```

I downloaded all the files from Hyak on to my desktop to see what this python script is and how its working. I have python2 installed on one of my Ubuntu terminals, so I tried the command Mary wrote to run the fasta2genotype.py script. It's an interactive python script that requires inputs that are not on the command line which explains why it was failing on Klone. 

```
(base) ctarpey@DESKTOP-CG1JP0S:/mnt/d/PacificCodWGS/Mary_PcodAnnotations/Migrate$ python2 ./fasta2genotype.py ./batch_7.fa ./migrate_loci_list.txt ./migrate_stacksIDs_finalgenepop.txt NA ./pcod-korea-migrate-input
Output type? [1] Migrate [2] Arlequin [3] DIYABC [4] LFMM [5] G-Phocs [6] Haplotype: 1
Loci to use? [1] Variable [2] All: 2
Filter for allele frequency? False alleles might bias data. [1] Yes [2] No: 2
Remove monomorphic loci? [1] Yes [2] No: 2
Filter for missing genotypes? These might bias data. [1] Yes [2] No: 2
Clip cut sites? These may bias data. [1] Yes [2] No: 2
Cataloging populations...
Cataloging loci...

```

I'm not sure if I used the same answers that Mary would have, but it appears to be running now. 
