## Documentation

Some additional documentation saved into Jupyter notebooks during data processing / analysis. These notebooks have been carried over from another repository. They are intended for record-keeping, and so may not be user-friendly. 

- Stacks - to cstacks: Final run of Stacks, from `process_radtags` to `cstacks`. Note that this notebook does show code to run through `populations`, but this was not the final code for the manuscript (see next notebook, pstacks to filtering).

- Stacks - pstacks to filtering: Final run of Stacks, `pstacks` through `populations` with additional filtering for minor allele frequencies, hardy-weinberg, and missing genotypes

- Explore Potential Contamination: Combine observed heterozygosity and ML Relate to identify potentially contaminated individuals that should be left out of the analysis

- Identify Outliers: Outlier loci identification with OutFLANK and Bayescan

- Outlier Alignment: Align outlier loci (OutFLANK, Bayescan) to the Atlantic cod genome and then explore annotations.

- Migrate Input Files: *not included in manuscript.* creation of input files for coalescence-based estimates of Ne using the program Migrate






<br>
<br>
<br>

*Version control. Contents from PCod-Korea-repo/notebooks, in order: Stacks batch 7 nb1; Stacks batch 8 - nb1; Batch 8 - MLRelate and Ho-verif; Batch 8 Outlier ID-verif; Batch 8 - Outlier alignment verif*